### PR TITLE
fix: pricing decimal formatting for tokenized cart

### PR DESCRIPTION
**Pull request by frosso**
Fixes tokenized cart for https://github.com/Automattic/woocommerce-payments/issues/8277

#### Changes proposed in this Pull Request

The amount displayed for zero-decimal currencies is wrong when using the PRBs w/ tokenized cart.
This is because Stripe has a concept of "zero decimal currency". Usually, we pass the PRBs a value expressed in cents. But for zero-decimal-currencies (like the Japanese Yen), the amount should be in Yen (they don't have Yen-cents).

Fixing.

Before:
Screenshot 2024-06-21 at 6 39 33 PM

After:
Screenshot 2024-06-21 at 6 22 45 PM

#### Testing instructions

- Enable the "Enable Cart-Token implementation for PRBs" flag in your dev tools (you'll need to update them)
- Ensure you have Google Pay/Apple Pay enabled in the merchant's settings
- Go to WooCommerce > Settings > General > Number of decimals, change the decimals from `2` to `0` (for example - you could also try `1`)
- Go to WooCommerce > Settings > General > Currency, select JPY
- Go to a product page
- Click on the GooglePay/ApplePay button
- The amounts displayed in the GooglePay/ApplePay dialog should be the same as the one displayed on the product page

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8998` or branch name `fix/tokenized-cart-zero-decimal-currency` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/tokenized-cart-zero-decimal-currency) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b86f9657ea29f6d1848d1b64af2cf709d3a9b7dc
- Build time: 2024-06-24 14:00:41 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		it( 'transforms the price if the currency is configured with one decimal', () => {
+			// with one decimal, `180` would mean `18.0`.
+			// But since Stripe expects the price to be in cents, the return value should be `1800`
+			expect( transformPrice( 180, { currency_minor_unit: 1 } ) ).toBe(
+				1800
+			);
+		} );
+
+		it( 'transforms the price if the currency is configured with two decimals', () => {
+			// with one decimal, `1800` would mean `18.00`.
```
Comment (timur27) [inline client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js]:
```suggestion
			// with two decimals, `1800` would mean `18.00`.
```

File: client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			// But since Stripe expects the price to be in cents, the return value should be `1800`
+			expect( transformPrice( 1800, { currency_minor_unit: 2 } ) ).toBe(
+				1800
+			);
+		} );
+
+		it( 'transforms the price if the currency a zero decimal currency (e.g.: Yen)', () => {
```
Comment (timur27) [inline client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js]: Sorry for nitpicking, I think we have a typo here  😺 
```suggestion
		it( 'transforms the price if the currency is a zero decimal currency (e.g.: Yen)', () => {
```

File: client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			global.wcpayPaymentRequestParams.checkout.currency_decimals = 0;
+			// with zero decimals, `18` would mean `18`.
+			// But since Stripe expects the price to be in the minimum currency amount, the return value should be `18`
```
Comment (timur27) [inline client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js]: It doesn't seem that we need this
```suggestion
```

**Review by timur27 — APPROVED** [review]
Looks good 👍 Left a few cosmetic non-blocking comments, and I also have one question: when using currencies with two decimal places (I tried with USD & Hungarian forint), no matter how many decimal points I set in the settings, Google Pay modal always displays the price with two decimal points. The price is correct so I don't think that this behavior is incorrect, but rather worth mentioning.

image

image

File: client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js (hunk 1 above)
Comment (frosso) [reply client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js]: ope

File: client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js (hunk 3 above)
Comment (frosso) [reply client/tokenized-payment-request/transformers/__tests__/wc-to-stripe.test.js]: I left it as a copy of the one before - but it reads better without that line :D

**Comment by frosso** [comment]
> I also have one question: when using currencies with two decimal places (I tried with USD & Hungarian forint), no matter how many decimal points I set in the settings, Google Pay modal always displays the price with two decimal points. The price is correct so I don't think that this behavior is incorrect, but rather worth mentioning.

Yeah, it's up to GooglePay (or Stripe?) to decide whether decimals are displayed or not. With these changes, I'm just trying to ensure the correct price is sent over :)

**Pull request merged**